File: android/src/main/java/com/emarsys/reactnative/utils/mappers/ProductMapper.kt
package com.emarsys.reactnative.utils.mappers

import com.emarsys.predict.api.model.Product
import com.emarsys.reactnative.utils.MapUtils.put
import com.emarsys.reactnative.utils.MapUtils.toStringMap
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap

object ProductMapper {

  fun toWritableMap(product: Product): WritableMap {
    val writableMap = Arguments.createMap();
    writableMap.put("productId", product.productId);
    writableMap.put("title", product.title);
    writableMap.put("linkUrl", product.linkUrl);
    writableMap.put("feature", product.feature);
    writableMap.put("cohort", product.cohort);
    writableMap.put("customFields", product.customFields);
    writableMap.put("imageUrl", product.imageUrl);
    writableMap.put("zoomImageUrl", product.zoomImageUrl);
    writableMap.put("categoryPath", product.categoryPath);
    writableMap.put("available", product.available);
    writableMap.put("productDescription", product.productDescription);
    writableMap.put("price", product.price);
    writableMap.put("msrp", product.msrp);
    writableMap.put("album", product.album);
    writableMap.put("actor", product.actor);
    writableMap.put("artist", product.artist);
    writableMap.put("author", product.author);
    writableMap.put("brand", product.brand);
    writableMap.put("year", product.year);

    return writableMap;
  }

  fun toModel(readableMap: ReadableMap?): Product? {
    if (readableMap == null) {
      return null
    }

    return Product(
      productId = readableMap.getString("productId")!!,
      title = readableMap.getString("title")!!,
      linkUrl = readableMap.getString("linkUrl")!!,
      feature = readableMap.getString("feature")!!,
      cohort = readableMap.getString("cohort")!!,
      customFields = readableMap.getMap("customFields")!!.toStringMap(),
      imageUrlString = readableMap.getString("imageUrl"), // imageUrl
      zoomImageUrlString = readableMap.getString("zoomImageUrl"), // zoomImageUrl
      categoryPath = readableMap.getString("categoryPath"),
      available = if (readableMap.hasKey("available") && !readableMap.isNull("available")) readableMap.getBoolean("available") else null,
      productDescription = readableMap.getString("productDescription"),
      price = if (readableMap.hasKey("price") && !readableMap.isNull("price")) readableMap.getDouble("price").toFloat() else null,
      msrp = if (readableMap.hasKey("msrp") && !readableMap.isNull("msrp")) readableMap.getDouble("msrp").toFloat() else null,
      album = readableMap.getString("album"),
      actor = readableMap.getString("actor"),
      artist = readableMap.getString("artist"),
      author = readableMap.getString("author"),
      brand = readableMap.getString("brand"),
      year = if (readableMap.hasKey("year") && !readableMap.isNull("year")) readableMap.getInt("year") else null
    )
  }

}

File: ios/utils/mappers/ProductMapper.m
#import "ProductMapper.h"
#import "EMSProduct+Emarsys.h"

@implementation ProductMapper

static Mapper _toDictionary = ^NSDictionary *(EMSProduct *product) {
  NSMutableDictionary<NSString *, NSObject *> *dictionary = [NSMutableDictionary dictionary];
  dictionary[@"productId"] = product.productId;
  dictionary[@"title"] = product.title;
  dictionary[@"linkUrl"] = product.linkUrl.absoluteString;
  dictionary[@"feature"] = product.feature;
  dictionary[@"cohort"] = product.cohort;
  dictionary[@"customFields"] = product.customFields;
  dictionary[@"imageUrl"] = product.imageUrl.absoluteString;
  dictionary[@"zoomImageUrl"] = product.zoomImageUrl.absoluteString;
  dictionary[@"categoryPath"] = product.categoryPath;
  dictionary[@"available"] = product.available;
  dictionary[@"productDescription"] = product.productDescription;
  dictionary[@"price"] = product.price;
  dictionary[@"msrp"] = product.msrp;
  dictionary[@"album"] = product.album;
  dictionary[@"actor"] = product.actor;
  dictionary[@"artist"] = product.artist;
  dictionary[@"author"] = product.author;
  dictionary[@"brand"] = product.brand;
  dictionary[@"year"] = product.year;

  return dictionary;
};

+ (Mapper)toDictionary {
  return _toDictionary;
}

+ (EMSProduct *)toModel:(NSDictionary *)dictionary {
  if (dictionary == nil) {
    return nil;
  }

  return [EMSProduct makeWithBuilder:^(EMSProductBuilder *builder) {
    [builder setRequiredFieldsWithProductId:dictionary[@"productId"]
      title:dictionary[@"title"]
      linkUrl:[NSURL URLWithString:dictionary[@"linkUrl"]]
      feature:dictionary[@"feature"]
      cohort:dictionary[@"cohort"]];
    [builder setCustomFields:dictionary[@"customFields"]];
    [builder setImageUrl:(dictionary[@"imageUrl"] != nil ? [NSURL URLWithString:dictionary[@"imageUrl"]] : nil)];
    [builder setZoomImageUrl:(dictionary[@"zoomImageUrl"] != nil ? [NSURL URLWithString:dictionary[@"zoomImageUrl"]] : nil)];
    [builder setCategoryPath:dictionary[@"categoryPath"]];
    [builder setAvailable:[NSNumber numberWithBool:dictionary[@"available"]]];
    [builder setProductDescription:dictionary[@"productDescription"]];
    [builder setPrice:dictionary[@"price"]];
    [builder setMsrp:dictionary[@"msrp"]];
    [builder setAlbum:dictionary[@"album"]];
    [builder setActor:dictionary[@"actor"]];
    [builder setArtist:dictionary[@"artist"]];
    [builder setAuthor:dictionary[@"author"]];
    [builder setBrand:dictionary[@"brand"]];
    [builder setYear:dictionary[@"year"]];
  }];
}

@end
